### ISSUE TITLE
Args parser

### ISSUE THREAD

**Issue by dvpfagan** [jmmcd/PonyGE2#25]
Might I suggest we move to using this https://docs.python.org/2/library/argparse.html#module-argparse

Dave

**Comment by jmmcd** [comment]
A comparison of several options: https://realpython.com/blog/python/comparing-python-command-line-parsing-libraries-argparse-docopt-click/

Brief HN discussion: https://news.ycombinator.com/item?id=10185290

Argparse (Python 3 docs): https://docs.python.org/3.7/library/argparse.html

Other options:

* Click http://click.pocoo.org/5/
* Docopt http://docopt.org


All of these would do a slightly better job of parsing our args, but they don't deal with the defaults + config file + command line args for us.

**Comment by mikefenton** [comment]
Stefan has accidentally committed a working version of a cleaner argument parser to the full default branch. You can check it out if you want to see how it works. Opinions/suggestions welcome!

**Comment by mikefenton** [comment]
Args parser is now formatted nicely and working correctly. Check it out at your leisure, we can discuss on Monday.

**Comment by jmmcd** [comment]
I think we used to have a check for non-existent command line options, but we don't anymore:

```
python ponyge.py --elephant

Start:	 2016-11-29 12:55:22.458737
(It runs happily!)
```

**Comment by mikefenton** [comment]
This is listed as a TODO. We parse non-existent command line options, we just don't do anything with them at the moment. The question is what do we want to do with them.

**Comment by mikefenton** [comment]
Check the first TODO in algorithm.parameters.set_params()

**Comment by dvpfagan** [comment]
I think being able to add unknown params at the command line is a terrible
idea. If you haven't added them to the dictionary or the params.txt then
tough luck. Throw an error saying parm not defined and move on with life :D

**Comment by t-h-e** [comment]
@dvpfagan shouldn't the same logic apply to the params file? If you haven't defined it in the dict/code, you should not be able to use it.

Either extra params in the param file and command line can be defined or no extra params can be defined at all.

**Comment by dvpfagan** [comment]
I'm all about just the params dict. :D  I vote to just use params dict :D

**Comment by mikefenton** [comment]
I'm leaning towards Stefan's point of view. If we allow people to add new key:value pairs to the params dictionary from a params file, why not from the command line too? If they make spelling mistakes and mess things up, then surely it's their fault. I don't think it's our responsibility to put preventative measures in place for such an event.

**Comment by jmmcd** [comment]
I don't mind too much which way it goes.

But if we aren't going to throw an error, we can still warn the user to catch spelling mistakes. Since ponyge.py is quite verbose (e.g. the warning about unit codons) we could print a warning to say "Options --elephant unrecognised by PonyGE2 parser; may be recognised by user code."

This would be a usability win as I found out this afternoon when passing `--search-loop` instead of `--search_loop` about 10 times in a row.

(BTW `--search-loop` is more common, see e.g. `man curl`.)

**Comment by mikefenton** [comment]
Sounds like a good compromise that fits with the rest of the PonyGE implementation.

**Comment by jmmcd** [comment]
Let's crash on mis-spelled args. User can use --extra-fitness-arguments for the fitness function. Otherwise new parameters they have to be added to the code.

**Comment by jmmcd** [comment]
Fixed 9d66639..cb64e8c
